### ISSUE TITLE
`mpileup` failed for large samples input (> 1000 BAM)

### ISSUE THREAD

**Issue by ydLiu-HIT** [samtools/samtools#1637]
#### Are you using the latest version of [samtools](https://github.com/samtools/samtools/releases/latest) and [HTSlib]
using samtools v1.15.1

 #### Please describe your environment.
Distributor ID: CentOS
Description:    CentOS Linux release 7.5.1804 (Core)
Release:        7.5.1804
Codename:       Core

#### Please specify the steps taken to generate the issue, the command you are running and the relevant output.

When I use `samtools mpileup` to generate pileups for large samples' BAM files( > 1000), it failed and output the error, that is "Could not load local index file" for one BAM. 

Then I test `mpileup` using one BAM file with a different number of copies. Firstly I copy the BAM 100,200,500 and 1000 times to the list of input BAM file `bam.list`, respectively, and run with the following command:
```
samtools mpileup -b bam.list -r chr22:20000000-20050000 -o results.pileup -d 100
```
with 100,200,500 and 1000 times copies，`mpileup` works fine. But when I increase the copies to 1200, 1500, 2000 times and run with the same command above, `mpileup` reports an error directly, as follows:
```
[E::hts_idx_load3] Could not load local index file '/home/user/liuyadong/joint/for_joint/HG002_benchmark/alignment.chr22.bam.bai'
[mpileup] fail to load index for /home/user/liuyadong/joint/for_joint/HG002_benchmark/alignment.chr22.bam
```
I want to know if `mpileup` has limitations on the number of input samples or some other reasons that caused the error.
Looking forward to your reply. Thanks!

**Comment by jkbonfield** [comment]
This may be an issue with the number of open file handles?  What does `ulimit -n` say?  If it's low, then you may be able to increase it, but that depends on the system hard limits.

I can't think what else would cause this failure.  You may also get some more information by doing something like `strace -o trace.out samtools mpileup -b bam.list -r chr22:20000000-20050000 -o results.pileup -d 100` and then looking for the failing filename somewhere near the end of the trace.out file, although if it takes a long while to fail this may be volumous and slow.

**Comment by ydLiu-HIT** [comment]
Sorry to bother you again. I have tried `ulimit -n` and the limited number of open file handles was set 100000, but `samtools mpileup` still crashed when the number of input BAM > 1020. And always report an error for the 1021th BAM as follow:
```
[E::hts_idx_load3] Could not load local index file '/home/user/liuyadong/joint/for_joint/HG002_benchmark/alignment.chr22.bam.bai'
[mpileup] fail to load index for /home/user/liuyadong/joint/for_joint/HG002_benchmark/alignment.chr22.bam
```
But I have checked the index was created and the BAM can be run correctly standalone.
So I was still confused about whether the error was caused by the limitation of `mpileup` or the system?

Looking forward to your reply.

**Comment by jkbonfield** [comment]
1020, with stdin, stdout, stderr and results.out bringing it to 1024, does seem uncannily like a file handle limit to me.  There aren't any obvious limitations in mpileup that stop after 1024 file handles.  I created 2000 files and ran your command.  With `ulimit -n 2003` I get the exact same error you saw, while `ulimit -n 2004` works, so it shows the 1024 limit is not in our code.

Are you batch submitting your job somewhere?  If so put a `ulimit -a` in the batch script to check the number you get reported on the command line matches the one used in the job environment.

Did you try using `strace` as suggested to see what the system error is on opening the file?    I did, with a small ulimit -n, and see this:

```
openat(AT_FDCWD, "_1012.bam.bai", O_RDONLY) = -1 EMFILE (Too many open files)
```

This implies we should be improving our error messages to pass on the system error to the user.
